### PR TITLE
Serializer: implement encoding and decoding of `Target`

### PR DESCRIPTION
**Pull request by var-const**

### CODE REVIEW

**Comment by var-const** [comment]
@wilhuff This is ready for an early review if you'd like. Tests pass except for the issue with limits.

**Comment by var-const** [comment]
One thing I don't understand is how to distinguish which failures in `Decode` methods should fail the `Reader`, and which should be hard asserts.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 1]
```diff
@@ -1004,6 +1102,278 @@ TEST_F(SerializerTest, DecodeMaybeDocWithoutFoundOrMissingSetShouldFail) {
...
+}
+
+TEST_F(SerializerTest, EncodesNullFilter) {
+  SCOPED_TRACE("EncodesNullFilter");
+  ExpectUnaryOperator(Value(nullptr),
+                      v1::StructuredQuery_UnaryFilter_Operator_IS_NULL);
+}
+
+TEST_F(SerializerTest, EncodesNanFilter) {
+  SCOPED_TRACE("EncodesNanFilter");
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: It seems like this is something we could just do in the fixture, using the method in https://github.com/firebase/firebase-ios-sdk/pull/3899/files#diff-9cd97ed5d34722d070c417dea0c7b689R132

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 2]
```diff
@@ -939,26 +970,388 @@ Query Serializer::DecodeQueryTarget(
...
+  google_firestore_v1_StructuredQuery_CompositeFilter& composite =
+      result.composite_filter;
+  composite.op =
+      google_firestore_v1_StructuredQuery_CompositeFilter_Operator_AND;
+
+  auto count = CheckedSize(converted_filters.size());
+  composite.filters_count = count;
+  composite.filters =
+      MakeArray<google_firestore_v1_StructuredQuery_Filter>(count);
+  for (pb_size_t i = 0; i != count; ++i) {
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: I did it mostly for consistency with Objective-C.

Agreed that it's inefficient, though there's no getting around of having to traverse the list of filters once beforehand to know how many field filters it contains (so that it is known how much size to allocate for the array). Took a stab at doing this in a more efficient manner, PTAL.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 3]
```diff
@@ -939,26 +970,388 @@ Query Serializer::DecodeQueryTarget(
...
+      return result.push_back(DecodeFieldFilter(reader, proto.field_filter));
+
+    default:
+      reader->Fail(StringFormat("Unrecognized Filter.which_filter_type %s",
+                                proto.which_filter_type));
+      return result;
+  }
+}
+
+google_firestore_v1_StructuredQuery_Filter Serializer::EncodeNonCompositeFilter(
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Hmm, I did this renaming intentionally -- sorry I forgot to call it out. I'm fine with reverting to the previous version of the name, although it seems that `NonComposite` captures the idea better than an exhaustive list of non-composite filters.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 4]
```diff
@@ -939,26 +970,388 @@ Query Serializer::DecodeQueryTarget(
...
+  }
+}
+
+google_firestore_v1_StructuredQuery_Order* Serializer::EncodeOrderBys(
+    const OrderByList& orders) const {
+  auto* result = MakeArray<google_firestore_v1_StructuredQuery_Order>(
+      CheckedSize(orders.size()));
+
+  int i = 0;
+  for (const OrderBy& order : orders) {
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: I'm certainly not against it. It should make it harder to forget to increment the loop variable.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 5]
```diff
@@ -939,26 +970,388 @@ Query Serializer::DecodeQueryTarget(
...
+  google_firestore_v1_Cursor result{};
+  result.before = bound.before();
+
+  auto count = CheckedSize(bound.position().size());
+  result.values_count = count;
+  result.values = MakeArray<google_firestore_v1_Value>(count);
+
+  int i = 0;
+  for (const FieldValue& field_value : bound.position()) {
+    result.values[i] = EncodeFieldValue(field_value);
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Thanks for spotting this!

'Weak' as in 'non-existent'. Added a new test `EncodeBounds`, verified it fails if `i` is not incremented.

**Comment by var-const** [comment]
Should be ready for full review now.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 6]
```diff
@@ -939,26 +977,398 @@ Query Serializer::DecodeQueryTarget(
...
+               std::move(start_at), std::move(end_at));
+}
+
+google_firestore_v1_StructuredQuery_Filter Serializer::EncodeFilters(
+    const FilterList& filters) const {
+  google_firestore_v1_StructuredQuery_Filter result{};
+
+  auto is_field_filter = [](const Filter& f) { return f.IsAFieldFilter(); };
+  auto first = std::find_if(filters.begin(), filters.end(), is_field_filter);
+  size_t filters_count =
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: This statement is sufficiently complicated that I had a hard time understanding what you were doing. Why not just:

```c++
size_t filters_count = absl::c_count_if(filters, is_field_filter);
if (filters_count == 1) {
  auto first = absl::c_find_if(filters, is_field_filter);
  // ...
}
```

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: `NonComposite`, while perhaps accurate, sounds off to me. `Singular`? `Single`?

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 7]
```diff
@@ -939,26 +977,398 @@ Query Serializer::DecodeQueryTarget(
...
+
+  return FieldFilter::Create(std::move(field_path), op, std::move(value));
+}
+
+Filter Serializer::DecodeUnaryFilter(
+    nanopb::Reader* reader,
+    const google_firestore_v1_StructuredQuery_UnaryFilter& unary) const {
+  if (unary.which_operand_type !=
+      google_firestore_v1_StructuredQuery_UnaryFilter_field_tag) {
+    reader->Fail(StringFormat("Unexpected UnaryFilter.which_operand_type: %s",
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: This seems like it should be an assertion since this is a violation of the precondition of this call.

Contrast with the failure in `DecodeCompositeFilter`, which I think is appropriate since nothing else is checking that the operator is AND.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 8]
```diff
@@ -939,26 +977,398 @@ Query Serializer::DecodeQueryTarget(
...
+
+  return result;
+}
+
+std::shared_ptr<Bound> Serializer::DecodeBound(
+    nanopb::Reader* reader, const google_firestore_v1_Cursor& cursor) const {
+  std::vector<FieldValue> index_components;
+  index_components.reserve(cursor.values_count);
+
+  for (pb_size_t i = 0; i != cursor.values_count; ++i) {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: More conventionally this is written `i < cursor.values_count`.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 9]
```diff
@@ -388,6 +437,55 @@ class SerializerTest : public ::testing::Test {
...
+    EXPECT_TRUE(msg_diff.Compare(proto, actual_proto)) << message_differences;
+  }
+
+  void ExpectDeserializationRoundTrip(const QueryData& model,
+                                      const v1::Target& proto) {
+    core::Query actual_model;
+    if (proto.has_documents()) {
+      actual_model = Decode<google_firestore_v1_Target_DocumentsTarget>(
+          google_firestore_v1_Target_DocumentsTarget_fields,
+          std::mem_fn(&Serializer::DecodeDocumentsTarget), proto.documents());
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: You could also use it for `IsAFieldFilter`, above.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 10]
```diff
@@ -388,6 +437,55 @@ class SerializerTest : public ::testing::Test {
...
+      actual_model = Decode<google_firestore_v1_Target_QueryTarget>(
+          google_firestore_v1_Target_QueryTarget_fields,
+          std::mem_fn(&Serializer::DecodeQueryTarget), proto.query());
+    }
+
+    EXPECT_EQ(model.query(), actual_model);
+  }
+
+  template <typename T>
+  ByteString Encode(const pb_field_t* fields, T&& nanopb_proto) {
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Reducing boilerplate in here would be lovely.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 11]
```diff
@@ -1004,6 +1102,278 @@ TEST_F(SerializerTest, DecodeMaybeDocWithoutFoundOrMissingSetShouldFail) {
...
+
+TEST_F(SerializerTest, EncodesMultipleFiltersOnDeeperCollections) {
+  core::Query q =
+      Query("rooms/1/messages/10/attachments")
+          .AddingFilter(Filter("prop", ">=", 42))
+          .AddingFilter(Filter("author", "==", "dimond"))
+          .AddingFilter(Filter("tags", "array_contains", "pending"));
+  QueryData model = CreateQueryData(std::move(q));
+
+  v1::Target proto;
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: We could reduce the duplication in here with a few helpers. Agreed that it's a lot of code but I don't see any way around it if we were to migrate these tests to just using nanopb directly.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Bump. This would save needing to write SCOPED_TRACE in every method in here.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 7 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Done, thanks.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 6 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: This is more efficient because it avoids traversing the container twice (remember that this rewrite came out of a discussion about performance). But you're right that it's overkill.

Changed to the way you suggested. Ironically, this is what I had originally when I first started work on this function (then I changed it to the Objective-C way to be more "consistent", and so on).

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: I think I'd rather keep as-is. The gain is less clear in that case, so the common argument "Everyone knows lambdas, but `mem_fn` is obscure magic" seems to go in favor of the current approach.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 12]
```diff
@@ -858,78 +851,123 @@ FieldTransform Serializer::DecodeFieldTransform(
...
+}
+
+Query Serializer::DecodeDocumentsTarget(
+    nanopb::Reader* reader,
+    const google_firestore_v1_Target_DocumentsTarget& proto) const {
+  if (proto.documents_count != 1) {
+    reader->Fail(
+        StringFormat("DocumentsTarget contained other than 1 document %s",
+                     proto.documents_count));
+    return Query::Invalid();
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Done -- let me know if I misunderstood your intent.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 1 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Looking at it now, it seems like it's incorrect -- `ScopedTrace` constructor takes the file name and line as its first two arguments, which the code in `AsyncTest` hardcodes to the same values? Am I missing something here?

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Excellent, I plan on refactoring this once we're done with the implementation.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 6 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: For what it's worth, my perspective on this is that it's usually worth the effort to remove extra copies but rarely worth the effort to absolutely minimize the number of reads. The former can frequently be expressed quite readably (as was done here) while the latter can lead to some magic-looking or low-level contortions.

As always, feel free to push back.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: If you don't think it's a slam dunk, it's not worth pursuing. What's here is fine.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 8 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Interesting perspective.

From my point of view less-than is better because it decouples loop termination from the step. It essentially declares for which values of the loop index the loop is valid without regard for how exactly we're stepping. In this case we're expressing that we should not go past the end of the array.

This has the effect of making the loop potentially more resilient to change. For example, should someone come along later and add a special case where the loop index is bumped an extra step to exclude the next value, we wouldn't need to specially handle the boundary condition where the last entry prompted that. Depending on the use case that may or may not be valid, but at least you're not going to find out about it with an infinite loop (or in this case a crash after reading past the end of mapped memory).

In short, I think it's reasonable to write these loops as if the loop variable may be increased by more than one precisely because someone may come along and make a change to do just that.

Note: I'm mostly sharing perspective here rather than advocating that you necessarily make the change. If you agree, great. If not, feel free to ignore.
